notebook/notebook/notebook.py

```diff
@@ -113,201 +143,334 @@ def start_pod(jupyter_token, image):
-    if not session.get('admin'):
-        return redirect(external_url_for('admin-login'))
-    delete_worker_pod(pod_name, svc_name)
-    return redirect(external_url_for('workers'))
-
-
-@app.route('/workers/delete-all-workers', methods=['POST'])
-def delete_all_workers():
-    if not session.get('admin'):
-        return redirect(external_url_for('admin-login'))
-    workers_and_svcs = get_all_workers()
-    for pod_name, svc_name in workers_and_svcs:
-        delete_worker_pod(pod_name, svc_name)
-    return redirect(external_url_for('workers'))
-
-
-def delete_worker_pod(pod_name, svc_name):
-    try:
-        k8s.delete_namespaced_pod(
-            pod_name,
-            'default',
-            kube.client.V1DeleteOptions(),
-            _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
-    except kube.client.rest.ApiException as e:
-        log.info(f'pod {pod_name} or associated service already deleted {e}')
+        label_selector=f"user_id={UNSAFE_user_id_transform(user_id)}",
+        timeout_seconds=KUBERNETES_TIMEOUT_IN_SECONDS).items
+
+    return marshall_json(pods, pod_paths), 200
+
+
+@app.route('/jupyter/<svc_name>', methods=['DELETE'])
+def delete_notebook(svc_name: str):
+    # TODO: Is it possible to have a falsy token value and get here?
+    user_id = request.headers.get("User")
+
+    if is_falsy_str(user_id) or is_falsy_str(svc_name):
+        return forbidden()
+
+    escp_user_id = UNSAFE_user_id_transform(request.headers.get("User"))
+
     try:
-        k8s.delete_namespaced_service(
-            svc_name,
-            'default',
-            kube.client.V1DeleteOptions(),
-            _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
-    except kube.client.rest.ApiException as e:
-        log.info(f'service {svc_name} (for pod {pod_name}) already deleted {e}')
+        svc = k8s.read_namespaced_service(svc_name, 'default')
 
+        if svc.metadata.labels['user_id'] != escp_user_id:
+            return forbidden()
 
-@app.route('/admin-login', methods=['GET'])
-def admin_login():
-    return render_template('admin-login.html',
-                           form_action_url=external_url_for('admin-login'))
+        del_svc(svc_name)
 
+        pods = k8s.list_namespaced_pod(
+            namespace='default',
+            label_selector=f"uuid={svc.metadata.labels['uuid']}",
+            timeout_seconds=KUBERNETES_TIMEOUT_IN_SECONDS).items
 
-@app.route('/admin-login', methods=['POST'])
-def admin_login_post():
-    if request.form['password'] != ADMIN_PASSWORD:
-        return '403 Forbidden', 403
-    session['admin'] = True
-    return redirect(external_url_for('workers'))
+        if len(pods) == 0:
+            log.error(
+                f"svc_name: {svc_name} pod_uuid: {svc.metadata.labels['uuid']}: pod not found for given uuid")
+            return '', 200
 
+        if len(pods) > 1:
+            log.error(
+                f"svc_name: {svc_name} pod_uuid: {svc.metadata.labels['uuid']}: pod_uuid is not unique")
```

Comment:
likewise to below, I'm inclined to make this an assert since it's not a normal error condition and it should definitely be a 500 ISE and we probably shouldn't delete the pods (they may contain useful debugging information.